backend/src/plugins/addons/server/services/bulk-operations/import-csv.ts

```diff
@@ -0,0 +1,271 @@
+         updated: 0,
+         collection: {
+            singularName: schema.info.singularName,
+            pluralName: schema.info.pluralName,
+         },
+         errors,
+      };
+   };
+
+interface ParseCSVResult {
+   records: Record<string, string>[];
+   errors: string[];
+}
+
+function parseCSV(csvString: string): Promise<ParseCSVResult> {
+   return new Promise((resolve) => {
+      const errors: string[] = [];
+      const records: Record<string, string>[] = [];
+      csv.parseString(csvString, { headers: true })
+         .on('error', (error) => {
+            if (error.message.includes('column header mismatch')) {
+               errors.push(
+                  `Error while reading the input file: column header mismatch, check that the file is correctly formatted. Usually this happens when you have a comma in a field value, and the value is not wrapped in quotes.`
+               );
+            } else {
+               errors.push(
+                  `Error while reading the input file: ${error.message}`
+               );
+            }
+            resolve({
+               records: [],
+               errors,
+            });
+         })
+         .on('data', async (record) => {
+            records.push(record);
+         })
+         .on('end', () => {
+            resolve({
+               records,
+               errors,
+            });
+         });
+   });
+}
+
+interface SaveRecordContext {
+   schema: ContentTypeSchema;
+   strapi: Strapi;
+}
+
+async function saveRecord(
+   record: Record<string, string>,
+   context: SaveRecordContext
+) {
+   const input = prepareRecordInput(record, context.schema);
+   const isNewRecord = isBlank(record.id);
+   if (isNewRecord) {
+      await strapi.entityService.create(context.schema.uid, {
+         data: input,
+      });
+   } else {
+      await strapi.entityService.update(context.schema.uid, record.id, {
+         data: input,
+      });
+   }
+}
+
+function prepareRecordInput(
+   record: Record<string, string>,
+   schema: ContentTypeSchema
+) {
+   const scalarAttributesNames = getScalarAttributes(schema);
+   const requiredComponentAttributesNames =
+      getRequiredComponentAttributes(schema);
+   const requiredDynamicZoneAttributesNames =
+      getRequiredDynamicZoneAttributes(schema);
+   const isNewRecord = isBlank(record.id);
+   const input: Record<string, unknown> = {};
+   for (const [name, value] of Object.entries(record)) {
+      if (scalarAttributesNames.includes(name)) {
+         input[name] = csvValueToAttributeValue(value, schema.attributes[name]);
+      }
+   }
+   if (isNewRecord) {
+      for (const name of requiredComponentAttributesNames) {
+         input[name] = {};
+      }
+      for (const name of requiredDynamicZoneAttributesNames) {
+         input[name] = [];
+      }
+   }
+   if (typeof record.published === 'string') {
+      if (isTrue(record.published)) {
+         input.publishedAt = new Date().toISOString();
```

Comment:
Should we set `updatedAt` here too, so that it matches with `publishedAt`? Probably doesn't matter that much, but the consistency might be nice for end users.

Reply:
I think `updatedAt` autoupdates whenever the record is updated 🤔